File: python/test/t_ProbabilitySimulationAlgorithm_std.py
#! /usr/bin/env python

from __future__ import print_function
import sys
import openturns as ot

ot.TESTPREAMBLE()


def progress(percent):
    sys.stderr.write('-- progress=' + str(percent) + '%\n')


def stop():
    sys.stderr.write('-- stop?\n')
    return False


# We create a numerical math function
myFunction = ot.SymbolicFunction(['E', 'F', 'L', 'I'], ['-F*L^3/(3*E*I)'])

dim = myFunction.getInputDimension()

# We create a normal distribution point of dimension 1
mean = [0.0] * dim
# E
mean[0] = 50.0
# F
mean[1] = 1.0
# L
mean[2] = 10.0
# I
mean[3] = 5.0
sigma = [1.0] * dim
R = ot.IdentityMatrix(dim)
myDistribution = ot.Normal(mean, sigma, R)

# We create a 'usual' RandomVector from the Distribution
vect = ot.RandomVector(myDistribution)

# We create a composite random vector
output = ot.CompositeRandomVector(myFunction, vect)

# We create an Event from this RandomVector
myEvent = ot.ThresholdEvent(output, ot.Less(), -3.0)

# Monte Carlo
experiments = [ot.MonteCarloExperiment()]
# Quasi Monte Carlo
experiments.append(ot.LowDiscrepancyExperiment())
# Randomized Quasi Monte Carlo
experiment = ot.LowDiscrepancyExperiment()
experiment.setRandomize(True)
experiments.append(experiment)
# Importance sampling
mean[0] = 4.99689645939288809018e+01
mean[1] = 1.84194175946153282375e+00
mean[2] = 1.04454036676956398821e+01
mean[3] = 4.66776215562709406726e+00
myImportance = ot.Normal(mean, sigma, R)
experiments.append(ot.ImportanceSamplingExperiment(myImportance))
# Randomized LHS
experiment = ot.LHSExperiment()
experiment.setAlwaysShuffle(True)
experiments.append(experiment)

for experiment in experiments:

    ot.RandomGenerator.SetSeed(0)

    myAlgo = ot.ProbabilitySimulationAlgorithm(myEvent, experiment)
    myAlgo.setMaximumOuterSampling(250)
    myAlgo.setBlockSize(4)
    myAlgo.setMaximumCoefficientOfVariation(0.1)

    print('algo=', myAlgo)

    # Perform the simulation
    myAlgo.run()

    # Stream out the result
    print('algo result=', myAlgo.getResult())
    print('probability distribution=',
          myAlgo.getResult().getProbabilityDistribution())

    # Use the standard deviation as a stoping rule
    experiment = ot.MonteCarloExperiment()
    myAlgo = ot.ProbabilitySimulationAlgorithm(myEvent, experiment)
    myAlgo.setMaximumOuterSampling(250)
    myAlgo.setBlockSize(4)
    myAlgo.setMaximumCoefficientOfVariation(0.0)
    myAlgo.setMaximumStandardDeviation(0.1)
    myAlgo.setProgressCallback(progress)
    myAlgo.setStopCallback(stop)

    print('algo=', myAlgo)

    # Perform the simulation
    myAlgo.run()

    # Stream out the result
    print('algo result=', myAlgo.getResult())
    print('probability distribution=',
          myAlgo.getResult().getProbabilityDistribution())

print('-' * 32)
ot.RandomGenerator.SetSeed(0)
description = ot.Description()
description.add('composite vector/comparison event')
dim = 2
distribution = ot.Normal(dim)
Xvector = ot.RandomVector(distribution)
f = ot.SymbolicFunction(['x0', 'x1'], ['x0+x1'])
Yvector = ot.CompositeRandomVector(f, Xvector)
s = 1.0
event1 = ot.ThresholdEvent(Yvector, ot.Greater(), s)
description.add('composite vector/domain event')
domain1D = ot.LevelSet(ot.SymbolicFunction(
    ['x0'], ['sin(x0)']), ot.LessOrEqual(), -0.5)
event2 = ot.DomainEvent(Yvector, domain1D)
description.add('composite vector/interval event')
interval = ot.Interval(0.5, 1.5)
event3 = ot.ThresholdEvent(Yvector, interval)
description.add('process/domain event')
Xprocess = ot.WhiteNoise(distribution, ot.RegularGrid(0.0, 0.1, 10))
domain2D = ot.LevelSet(
    ot.SymbolicFunction(['x0', 'x1'], ['(x0-1)^2+x1^2']), ot.LessOrEqual(), 1.0)
event4 = ot.ProcessEvent(Xprocess, domain2D)
all_events = [event1, event2, event3, event4]
for i, event in enumerate(all_events):
    print(description[i])
    if event.isComposite():
        experiment = ot.MonteCarloExperiment()
        myAlgo = ot.ProbabilitySimulationAlgorithm(event, experiment)
    else:
        myAlgo = ot.ProbabilitySimulationAlgorithm(event)
    myAlgo.setMaximumOuterSampling(250)
    myAlgo.setBlockSize(4)
    myAlgo.setMaximumCoefficientOfVariation(0.1)
    myAlgo.run()
    print('MonteCarlo result=', myAlgo.getResult())
    print('probability distribution=',
          myAlgo.getResult().getProbabilityDistribution())

File: lib/test/t_ProbabilitySimulationAlgorithm_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of ProbabilitySimulationAlgorithm class
 *
 *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

static Bool stop(void * )
{
  std::cerr << "-- stop?" << std::endl;
  return false;
}

static void progress(Scalar percent, void * )
{
  std::cerr << "-- progress=" << percent << "%" << std::endl;
}

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);


  try
  {
    /* We create a numerical math function */
    Description input(4);
    input[0] = "E";
    input[1] = "F";
    input[2] = "L";
    input[3] = "I";
    SymbolicFunction myFunction(input, Description(1, "-F*L^3/(3*E*I)"));

    UnsignedInteger dim = myFunction.getInputDimension();
    /* We create a normal distribution point of dimension 1 */
    Point mean(dim, 0.0);
    mean[0] = 50.0; // E
    mean[1] =  1.0; // F
    mean[2] = 10.0; // L
    mean[3] =  5.0; // I
    Point sigma(dim, 1.0);
    IdentityMatrix R(dim);
    Normal myDistribution(mean, sigma, R);

    /* We create a 'usual' RandomVector from the Distribution */
    RandomVector vect(myDistribution);

    /* We create a composite random vector */
    CompositeRandomVector output(myFunction, vect);

    /* We create an Event from this RandomVector */
    ThresholdEvent myEvent(output, Less(), -3.0);

    Collection<WeightedExperiment> experiments(1, MonteCarloExperiment());
    // qmc
    experiments.add(LowDiscrepancyExperiment());
    // importance sampling
    mean[0] = 4.99689645939288809018e+01;
    mean[1] = 1.84194175946153282375e+00;
    mean[2] = 1.04454036676956398821e+01;
    mean[3] = 4.66776215562709406726e+00;
    Normal myImportance(mean, sigma, R);
    experiments.add(ImportanceSamplingExperiment(myImportance));
    // randomized qmc
    LowDiscrepancyExperiment lde;
    lde.setRandomize(true);
    experiments.add(lde);
    // randomized lhs
    LHSExperiment lhse;
    lhse.setAlwaysShuffle(true);
    experiments.add(lhse);

    for (UnsignedInteger i = 0; i < experiments.getSize(); ++ i)
    {
      RandomGenerator::SetSeed(0);

      /* We create a Monte Carlo algorithm */
      ProbabilitySimulationAlgorithm myAlgo(myEvent, experiments[i]);
      myAlgo.setMaximumOuterSampling(250);
      myAlgo.setBlockSize(4);
      myAlgo.setMaximumCoefficientOfVariation(0.1);
      myAlgo.setProgressCallback(&progress);
      myAlgo.setStopCallback(&stop);

      fullprint << "algo=" << myAlgo << std::endl;

      /* Perform the simulation */
      myAlgo.run();

      /* Stream out the result */
      fullprint << "algo result=" << myAlgo.getResult() << std::endl;
      fullprint << "Confidence length at level 99%=" << myAlgo.getResult().getConfidenceLength(0.99) << std::endl;
      fullprint << "Confidence length at level 80%=" << myAlgo.getResult().getConfidenceLength(0.8) << std::endl;
      fullprint << "Probability distribution=" << myAlgo.getResult().getProbabilityDistribution() << std::endl;

      /* Use the standard deviation as a stoping rule */
      myAlgo = ProbabilitySimulationAlgorithm(myEvent, experiments[i]);
      myAlgo.setMaximumOuterSampling(250);
      myAlgo.setBlockSize(4);
      myAlgo.setMaximumCoefficientOfVariation(0.0);
      myAlgo.setMaximumStandardDeviation(0.1);

      fullprint << "algo=" << myAlgo << std::endl;

      /* Perform the simulation */
      myAlgo.run();

      /* Stream out the result */
      fullprint << "algo result=" << myAlgo.getResult() << std::endl;
      fullprint << "Probability distribution=" << myAlgo.getResult().getProbabilityDistribution() << std::endl;
    }
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
